x-pack/plugins/ml/public/explorer/explorer.js

```diff
@@ -52,48 +114,782 @@ function mapSwimlaneOptionsToEuiOptions(options) {
+      this.setState(
+        { swimlaneViewByFieldName: '' },
+        () => {
+          // Unique influencers for the selected job(s).
+          let viewByOptions = [];
+
+          const selectedJobIds = getSelectedJobIds(this.props.jobs);
+          const fieldsByJob = { '*': [] };
+          _.each(mlJobService.jobs, (job) => {
+            // Add the list of distinct by, over, partition and influencer fields for each job.
+            let fieldsForJob = [];
+
+            const analysisConfig = job.analysis_config;
+            const detectors = analysisConfig.detectors || [];
+            _.each(detectors, (detector) => {
+              if (_.has(detector, 'partition_field_name')) {
+                fieldsForJob.push(detector.partition_field_name);
+              }
+              if (_.has(detector, 'over_field_name')) {
+                fieldsForJob.push(detector.over_field_name);
+              }
+              // For jobs with by and over fields, don't add the 'by' field as this
+              // field will only be added to the top-level fields for record type results
+              // if it also an influencer over the bucket.
+              if (_.has(detector, 'by_field_name') && !(_.has(detector, 'over_field_name'))) {
+                fieldsForJob.push(detector.by_field_name);
+              }
+            });
+
+            const influencers = analysisConfig.influencers || [];
+            fieldsForJob = fieldsForJob.concat(influencers);
+            if (selectedJobIds.indexOf(job.job_id) !== -1) {
+              viewByOptions = viewByOptions.concat(influencers);
+            }
+
+            fieldsByJob[job.job_id] = _.uniq(fieldsForJob);
+            fieldsByJob['*'] = _.union(fieldsByJob['*'], fieldsByJob[job.job_id]);
+          });
+
+          // Currently unused but may be used if add in view by detector.
+          // $scope.fieldsByJob = fieldsByJob;
+          viewByOptions = _.chain(viewByOptions).uniq().sortBy(fieldName => fieldName.toLowerCase()).value();
+          viewByOptions.push(VIEW_BY_JOB_LABEL);
+          const viewBySwimlaneOptions = viewByOptions;
+
+          let swimlaneViewByFieldName = null;
+
+          if (this.props.appState.mlExplorerSwimlane.viewBy !== undefined &&
+            this.state.viewBySwimlaneOptions.indexOf(this.props.appState.mlExplorerSwimlane.viewBy) !== -1) {
+            // Set the swimlane viewBy to that stored in the state (URL) if set.
+            swimlaneViewByFieldName = this.props.appState.mlExplorerSwimlane.viewBy;
+          } else {
+            if (selectedJobIds.length > 1) {
+              // If more than one job selected, default to job ID.
+              swimlaneViewByFieldName = VIEW_BY_JOB_LABEL;
+            } else {
+              // For a single job, default to the first partition, over,
+              // by or influencer field of the first selected job.
+              const firstSelectedJob = _.find(mlJobService.jobs, (job) => {
```

Comment:
More lodash which could be replaced by `Array.find`? Plus more use of `_.each` below

Reply:
Done in 7ec1bfedaf.